util/topgen.py

```diff
@@ -543,106 +659,103 @@ def generate_flash(topcfg: Dict[str, object], out_path: Path) -> None:
-                mapping_path,
-                if_name,
-                name_to_block[m['type']])
-            m['racl_mappings'][if_name] = {
-                'racl_group': racl_group,
-                'register_mapping': parsed_register_mapping,
-                'window_mapping': parsed_window_mapping,
-            }
+    for m in topcfg["module"]:
+        racl_mappings = m.get("racl_mappings", {})
+        for if_name, mapping in racl_mappings.items():
+            racl_group = racl_mappings[if_name]["racl_group"]
             num_subscribing_ips[racl_group] += 1
 
-    log.info('Generating RACL Control IP with ipgen')
-    topname = topcfg['name']
-
-    # Get the AC Range Check instance
-    racl_ctrl = lib.find_module(topcfg['module'], 'racl_ctrl')
-
-    if len(topcfg['racl']['policies']) == 1:
-        # If there is only one set of policies, take the first one
-        policies = list(topcfg['racl']['policies'].values())[0]
-    else:
-        # More than one policy, we need to find the matching set of policies
-        racl_group = racl_ctrl['racl_group']
-        policies = topcfg['racl']['policies'][racl_group]
-
-    params = {
-        'module_instance_name': racl_ctrl['type'],
-        'nr_role_bits': topcfg['racl']['nr_role_bits'],
-        'nr_ctn_uid_bits': topcfg['racl']['nr_ctn_uid_bits'],
-        'nr_policies': len(policies),
+    return {
+        "module_instance_name": racl_ctrl["type"],
+        "nr_role_bits": topcfg["racl"]["nr_role_bits"],
+        "nr_ctn_uid_bits": topcfg["racl"]["nr_ctn_uid_bits"],
+        "nr_policies": len(policies),
         'nr_subscribing_ips': num_subscribing_ips[racl_group],
-        'policies': policies
+        "policies": policies
     }
 
-    ipgen_render("racl_ctrl", topname, params, out_path)
-
 
-def generate_top_only(top_only_dict: Dict[str, bool], out_path: Path,
-                      top_name: str, alt_hjson_path: str) -> None:
+# Generate RACL collateral
+def generate_racl(topcfg: Dict[str, object], module: Dict[str, object],
+                  out_path: Path) -> None:
+    # Not all tops use RACL
+    if "racl_config" not in topcfg:
+        raise ValueError(
+            "There is a racl_ctrl module but no 'racl_config' in top config")
+    log.info("Generating RACL Control IP with ipgen")
+    params = _get_racl_params(topcfg)
+    generate_ipgen(topcfg, module, params, out_path)
+
+
+# def generate_top_only(top_only_dict: Dict[str, bool], out_path: Path,
+def generate_top_only(top_only_dict: List[str], out_path: Path, top_name: str,
+                      alt_hjson_path: str) -> None:
+    """Generate the regfile for top_only IPs."""
     log.info("Generating top only modules")
 
-    for ip, reggen_only in top_only_dict.items():
+    for ip in top_only_dict:
         ip_out_path = out_path / "ip" / ip
-        if reggen_only and alt_hjson_path:
-            hjson_path = Path(alt_hjson_path)
-        else:
-            hjson_path = ip_out_path / "data" / f"{ip}.hjson"
```

Comment:
Are we sure we should be killing this feature? I don't know who uses it, but there doesn't seem to be a great reason to remove it.

That said... does it even work anymore? It looks like this was broken by https://github.com/lowRISC/opentitan/pull/25129, undoing the support that was completed for https://github.com/lowRISC/opentitan/issues/8207.

In a world where the tops in the open-source repo are reference tops and Nuvoton's top hjson is out-of-tree, I guess we don't really need that argument anymore, do we? If an integrator wanted to work with partners using a sort of "redacted" version of their internal top, I guess they could provide a separate toplevel.hjson that has the proprietary IPs replaced with shareable models.

Reply:
This flag is not even sensible: imagine we had multiple reggens, then they would all use the same hjson file? I vaguely recall talking to someone else about this and agreeing we don't need it anyway.